cmd/vspd/spentticket.go

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2023 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.package main
+
+package main
+
+import (
+	"encoding/hex"
+	"fmt"
+
+	"github.com/decred/dcrd/blockchain/stake/v5"
+	"github.com/decred/dcrd/rpc/jsonrpc/types/v4"
+	"github.com/decred/dcrd/txscript/v4/stdaddr"
+	"github.com/decred/dcrd/wire"
+	"github.com/decred/vspd/database"
+)
+
+type spentTicket struct {
+	dbTicket     database.Ticket
+	expiryHeight int64
+	heightSpent  int64
+	spendingTx   *wire.MsgTx
+}
+
+func (s *spentTicket) voted() bool {
+	return stake.IsSSGen(s.spendingTx)
+}
+
+// findSpentTickets attempts to find transactions that vote/revoke the provided
+// tickets by matching the payment script of the tickets commitment address
+// against the block filters of the mainchain blocks between the provided start
+// block and the current best block. Returns any found spent tickets and the
+// height of the most recent scanned block.
+func (v *vspd) findSpentTickets(toCheck database.TicketList, startHeight int64) (map[string]spentTicket, int64, error) {
+	params := v.cfg.netParams
+
+	dcrdClient, _, err := v.dcrd.Client()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	bestBlock, err := dcrdClient.GetBestBlockHeader()
+	if err != nil {
+		return nil, 0, fmt.Errorf("dcrd.GetBestBlockHeader error: %w", err)
+	}
+	endHeight := int64(bestBlock.Height)
+
+	if startHeight > endHeight {
+		return nil, 0, fmt.Errorf("start height %d greater than best block height %d",
+			startHeight, endHeight)
+	}
+
+	numBlocks := int(1 + endHeight - startHeight)
+
+	// Only log if checking a larger number of blocks to avoid spam.
+	if numBlocks > 5 {
+		v.log.Debugf("Scanning %d blocks for %s",
+			numBlocks, pluralize(len(toCheck), "spent ticket"))
+	}
+
+	// Get commitment address payment script for each ticket.
+	type ticketTuple struct {
+		dbTicket database.Ticket
+		pkScript []byte
+	}
+
+	tickets := make(map[string]ticketTuple)
+	for _, ticket := range toCheck {
+		parsedAddr, err := stdaddr.DecodeAddress(ticket.CommitmentAddress, params)
+		if err != nil {
+			return nil, 0, err
+		}
+		_, script := parsedAddr.PaymentScript()
+
+		tickets[ticket.Hash] = ticketTuple{
+			dbTicket: ticket,
+			pkScript: script,
+		}
+	}
+
+	spent := make(map[string]spentTicket)
+
+	for iHeight := startHeight; iHeight <= endHeight; iHeight++ {
+		iHash, err := dcrdClient.GetBlockHash(iHeight)
+		if err != nil {
+			return nil, 0, err
+		}
+
+		iBlock, err := dcrdClient.GetBlockVerbose(iHash)
```

Comment:
I imagine you're not super concerned with performance here, but it would be a lot more efficient to avoid getting verbose results and instead just get the block itself (by setting the two verbose flags in the RPC).

With both flags false, you just get the raw hex-encoded block.  So you take that and decode it in a `wire.MsgBlock`.  I know you're not using dcrd's RPC client, but relevant code [here](https://github.com/decred/dcrd/blob/0a35adc02120805f5185cc0a53eb2674711e3013/rpcclient/chain.go#L73-L92).

That way you wouldn't need to round trip through additional hex encoding nor incur the overhead the extra things verbose provides that you don't need here.